File: frontend/features/attendance/types/attendance.ts
export interface Attendance {
  id: string;
  practice_schedule_id: string;
  user_id: string;
  status: 'present' | 'absent' | 'late' | 'no_show';
  notes?: string;
  available_from?: string; // 参加開始時刻（HH:MM形式）
  available_to?: string; // 参加終了時刻（HH:MM形式）
  created_at: string;
  updated_at: string;
  created_by: string;
  updated_by: string;
  user_name?: string; // ユーザー名
  user_email?: string; // ユーザーメールアドレス
}

export interface AttendanceCreate {
  practice_schedule_id: string;
  user_id: string;
  status: 'present' | 'absent' | 'late' | 'no_show';
  notes?: string;
  available_from?: string; // 参加開始時刻（HH:MM形式）
  available_to?: string; // 参加終了時刻（HH:MM形式）
}

export interface AttendanceUpdate {
  status?: 'present' | 'absent' | 'late' | 'no_show';
  notes?: string;
  available_from?: string; // 参加開始時刻（HH:MM形式）
  available_to?: string; // 参加終了時刻（HH:MM形式）
}

export interface AttendanceResponse {
  id: string;
  practice_schedule_id: string;
  user_id: string;
  status: 'present' | 'absent' | 'late' | 'no_show';
  notes?: string;
  available_from?: string; // 参加開始時刻（HH:MM形式）
  available_to?: string; // 参加終了時刻（HH:MM形式）
  created_at: string;
  updated_at: string;
  created_by: string;
  updated_by: string;
}

export interface AttendanceSummary {
  practice_schedule_id: string;
  practice_date: string;
  practice_time: string;
  venue_name: string;
  total_users: number;
  present_count: number;
  absent_count: number;
  late_count: number;
  no_show_count: number;
}

export interface UserAttendanceHistory {
  user_id: string;
  user_name: string;
  total_practices: number;
  present_count: number;
  absent_count: number;
  late_count: number;
  no_show_count: number;
  attendance_rate: number;
}

export interface PracticeSchedule {
  id: string;
  schedule_date: string;
  start_time: string;
  end_time: string;
  division_count: number;
  title?: string;
  description?: string;
  schedule_type?: string;
  status?: string;
  created_at?: string;
  updated_at?: string;
  created_by?: string;
  updated_by?: string;
  venues?: {
    id: string;
    name: string;
    campus: string;
  }[];
}

export interface AttendanceFormData {
  practice_schedule_id: string;
  status: 'present' | 'absent' | 'late' | 'no_show';
  notes: string;
  available_from?: string; // 参加開始時刻（HH:MM形式）
  available_to?: string; // 参加終了時刻（HH:MM形式）
}

export interface AttendanceFormErrors {
  practice_schedule_id?: string;
  status?: string;
  notes?: string;
  available_from?: string;
  available_to?: string;
}

File: backend/app/schemas/attendance.py
from datetime import datetime, time
from uuid import UUID
from enum import Enum

from pydantic import BaseModel, ConfigDict


class AttendanceStatus(str, Enum):
    """出席状況の列挙型"""
    PRESENT = "present"      # 出席
    ABSENT = "absent"        # 欠席
    LATE = "late"           # 遅刻
    NO_SHOW = "no_show"     # 無断欠席
    UNDECIDED = "undecided"  # 未決定（後方互換性のため）


class AttendanceBase(BaseModel):
    """出欠記録の基本情報"""

    practice_schedule_id: UUID
    user_id: UUID
    status: AttendanceStatus
    notes: str | None = None
    available_from: time | None = None
    available_to: time | None = None


class AttendanceCreate(AttendanceBase):
    """出欠記録作成用スキーマ"""
    pass


class AttendanceUpdate(BaseModel):
    """出欠記録更新用スキーマ"""

    status: AttendanceStatus | None = None
    notes: str | None = None
    available_from: time | None = None
    available_to: time | None = None


class AttendanceResponse(AttendanceBase):
    """出欠記録レスポンス用スキーマ"""

    id: UUID
    created_at: datetime | None = None
    updated_at: datetime | None = None
    created_by: UUID | None = None
    updated_by: UUID | None = None
    user_name: str | None = None  # フルネーム（last_name_kanji first_name_kanji）
    user_email: str | None = None  # ユーザーのメールアドレス
    user_year: int | None = None  # ユーザーの学年

    model_config = ConfigDict(from_attributes=True)


class AttendanceSummary(BaseModel):
    """練習別出欠サマリー用スキーマ"""
    
    practice_schedule_id: UUID
    schedule_date: datetime
    description: str | None = None
    venue_name: str | None = None
    total_people: int
    present_count: int
    absent_count: int
    late_count: int
    no_show_count: int
    attendance_rate: float

    model_config = ConfigDict(from_attributes=True)


class UserAttendanceHistory(BaseModel):
    """ユーザー別出欠履歴用スキーマ"""

    user_id: UUID
    email: str
    first_name_kanji: str | None = None
    last_name_kanji: str | None = None
    student_id: str | None = None
    attendance_status: AttendanceStatus | None = None
    available_from: time | None = None
    available_to: time | None = None
    schedule_date: datetime | None = None
    description: str | None = None
    venue_name: str | None = None
    notes: str | None = None

    model_config = ConfigDict(from_attributes=True)


